File: scripts/visual-audit/capture.py
#!/usr/bin/env python3
"""Verificación de consistencia visual (design language).

Recorre rutas principales autenticadas y guarda una captura por ruta para
revisar side-by-side que headers, filtros, badges y KPIs sigan homologados
tras las Olas 1–7.

Uso:
    python scripts/visual-audit/capture.py [--base http://localhost:8080] [--out ./visual-snapshots]

Credenciales por defecto: mem://reference/audit-login (override con
AUDIT_EMAIL / AUDIT_PASSWORD).
"""
from __future__ import annotations

import argparse
import asyncio
import json
import os
from datetime import datetime, timezone
from pathlib import Path

from playwright.async_api import async_playwright

ROUTES = [
    # Núcleo operativo
    ("01-inicio", "/inicio"),
    ("02-operaciones", "/operaciones"),
    ("03-embarques", "/embarques"),
    ("04-cotizaciones", "/cotizaciones"),
    ("05-proformas", "/proformas"),
    ("06-facturacion", "/facturacion"),
    # Financiero
    ("10-compras", "/compras"),
    ("11-cxp", "/cxp"),
    ("12-cartera", "/cartera"),
    ("13-tesoreria", "/tesoreria"),
    ("14-profit-dashboard", "/profit/dashboard"),
    ("15-profit-proyeccion", "/profit/proyeccion"),
    # Catálogos / CRM
    ("20-clientes", "/clientes"),
    ("21-proveedores", "/proveedores"),
    # Costeo
    ("30-costeo-tarifas", "/costeo/tarifas"),
    ("31-costeo-rutas", "/costeo/rutas"),
    ("32-costeo-navieras", "/costeo/navieras"),
    ("33-costeo-agentes", "/costeo/agentes"),
    # Reportes
    ("40-reportes-rentabilidad", "/reportes/rentabilidad"),
    ("41-reportes-cierre", "/reportes/cierre-mensual"),
    # Admin plataforma
    ("50-admin", "/admin"),
    ("51-admin-organizaciones", "/admin/organizaciones"),
    ("52-admin-auditoria", "/admin/auditoria"),
]

VIEWPORT = {"width": 1440, "height": 900}


async def login(page, base: str, email: str, password: str) -> None:
    await page.goto(f"{base}/login", wait_until="domcontentloaded")
    await page.locator("#email").fill(email)
    await page.locator("#password").fill(password)
    await page.get_by_role("button", name="Iniciar sesión").click()
    await page.wait_for_url(lambda url: "/login" not in url, timeout=25_000)


async def capture(page, base: str, slug: str, path: str, out_dir: Path) -> dict:
    errors: list[str] = []

    def on_console(msg):
        if msg.type == "error":
            errors.append(msg.text[:200])

    page.on("console", on_console)
    try:
        try:
            await page.goto(f"{base}{path}", wait_until="networkidle", timeout=20_000)
        except Exception:
            await page.wait_for_timeout(500)
        await page.wait_for_timeout(700)
        file_path = out_dir / f"{slug}.png"
        await page.screenshot(path=str(file_path))
        return {"slug": slug, "path": path, "file": file_path.name, "errors": len(errors), "errorSamples": errors[:3]}
    finally:
        page.remove_listener("console", on_console)


async def run(base: str, out: Path, email: str, password: str) -> None:
    out.mkdir(parents=True, exist_ok=True)
    async with async_playwright() as pw:
        browser = await pw.chromium.launch(headless=True)
        ctx = await browser.new_context(viewport=VIEWPORT, locale="es-MX")
        page = await ctx.new_page()
        try:
            await login(page, base, email, password)
            results = []
            for slug, path in ROUTES:
                print(f"→ {path}")
                results.append(await capture(page, base, slug, path, out))
        finally:
            await browser.close()

    report = {
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "base": base,
        "viewport": VIEWPORT,
        "routes": results,
    }
    (out / "report.json").write_text(json.dumps(report, indent=2, ensure_ascii=False))

    lines = [
        "# Reporte de consistencia visual",
        "",
        f"Generado: {report['generatedAt']}  ·  Base: {base}  ·  Viewport: {VIEWPORT['width']}×{VIEWPORT['height']}",
        "",
        "| # | Ruta | Errores consola | Captura |",
        "|---|------|-----------------|---------|",
    ]
    for i, r in enumerate(results, 1):
        lines.append(f"| {i} | `{r['path']}` | {r['errors']} | `{r['file']}` |")
    (out / "REPORT.md").write_text("\n".join(lines))
    print(f"\n✅ {len(results)} capturas en {out}")


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--base", default=os.environ.get("AUDIT_BASE_URL", "http://localhost:8080"))
    ap.add_argument("--out", default="./visual-snapshots")
    args = ap.parse_args()

    email = os.environ.get("AUDIT_EMAIL", "hector@lopezbenavides.com")
    password = os.environ.get("AUDIT_PASSWORD", "1234567890")

    asyncio.run(run(args.base, Path(args.out).resolve(), email, password))


if __name__ == "__main__":
    main()

File: scripts/visual-audit/capture.mjs
#!/usr/bin/env node
/**
 * Verificación de consistencia visual (design language).
 *
 * Recorre las rutas principales autenticadas y guarda una captura por ruta
 * para revisar side-by-side que headers, filtros, badges y KPIs estén
 * homologados tras las Olas 1–7.
 *
 * Uso:
 *   node scripts/visual-audit/capture.mjs [--base=http://localhost:8080] [--out=./visual-snapshots]
 *
 * Requiere: `playwright` con Chromium. En sandbox de Lovable ya viene instalado
 * (PLAYWRIGHT_BROWSERS_PATH pre-seteado). Fuera del sandbox correr:
 *   bunx playwright install chromium
 *
 * Credenciales: usa AUDIT_EMAIL / AUDIT_PASSWORD (ver mem://reference/audit-login).
 */
import { chromium } from "playwright";
import { mkdir } from "node:fs/promises";
import { existsSync, writeFileSync } from "node:fs";
import path from "node:path";

const argv = Object.fromEntries(
  process.argv.slice(2).map((a) => {
    const [k, v] = a.replace(/^--/, "").split("=");
    return [k, v ?? true];
  }),
);

const BASE = argv.base ?? process.env.AUDIT_BASE_URL ?? "http://localhost:8080";
const OUT = path.resolve(argv.out ?? "./visual-snapshots");
const EMAIL = process.env.AUDIT_EMAIL ?? "hector@lopezbenavides.com";
const PASSWORD = process.env.AUDIT_PASSWORD ?? "1234567890";

/**
 * Rutas críticas por dominio funcional. Un slug estable por ruta se usa como
 * nombre de archivo para poder comparar entre corridas (baseline vs actual).
 */
const ROUTES = [
  // Núcleo operativo
  { slug: "01-inicio", path: "/inicio" },
  { slug: "02-operaciones", path: "/operaciones" },
  { slug: "03-embarques", path: "/embarques" },
  { slug: "04-cotizaciones", path: "/cotizaciones" },
  { slug: "05-proformas", path: "/proformas" },
  { slug: "06-facturacion", path: "/facturacion" },

  // Financiero
  { slug: "10-compras", path: "/compras" },
  { slug: "11-cxp", path: "/cxp" },
  { slug: "12-cartera", path: "/cartera" },
  { slug: "13-tesoreria", path: "/tesoreria" },
  { slug: "14-profit-dashboard", path: "/profit/dashboard" },
  { slug: "15-profit-proyeccion", path: "/profit/proyeccion" },

  // Catálogos / CRM
  { slug: "20-clientes", path: "/clientes" },
  { slug: "21-proveedores", path: "/proveedores" },

  // Costeo
  { slug: "30-costeo-tarifas", path: "/costeo/tarifas" },
  { slug: "31-costeo-rutas", path: "/costeo/rutas" },
  { slug: "32-costeo-navieras", path: "/costeo/navieras" },
  { slug: "33-costeo-agentes", path: "/costeo/agentes" },

  // Reportes
  { slug: "40-reportes-rentabilidad", path: "/reportes/rentabilidad" },
  { slug: "41-reportes-cierre", path: "/reportes/cierre-mensual" },

  // Admin plataforma
  { slug: "50-admin", path: "/admin" },
  { slug: "51-admin-organizaciones", path: "/admin/organizaciones" },
  { slug: "52-admin-auditoria", path: "/admin/auditoria" },
];

const VIEWPORT = { width: 1440, height: 900 };

async function login(page) {
  await page.goto(`${BASE}/login`, { waitUntil: "domcontentloaded" });
  await page.getByLabel("Email").fill(EMAIL);
  await page.getByLabel("Contraseña").fill(PASSWORD);
  await Promise.all([
    page.waitForURL((u) => !u.pathname.startsWith("/login"), { timeout: 20_000 }),
    page.getByRole("button", { name: /iniciar sesión/i }).click(),
  ]);
}

async function capture(page, route) {
  const url = `${BASE}${route.path}`;
  const errors = [];
  const onErr = (msg) => {
    if (msg.type() === "error") errors.push(msg.text());
  };
  page.on("console", onErr);
  try {
    await page.goto(url, { waitUntil: "networkidle", timeout: 20_000 });
  } catch (e) {
    // Continuar aunque no llegue a networkidle: capturamos lo que haya.
    await page.waitForTimeout(500);
  }
  await page.waitForTimeout(600); // deja asentar animaciones y skeletons
  const file = path.join(OUT, `${route.slug}.png`);
  await page.screenshot({ path: file, fullPage: false });
  page.off("console", onErr);
  return { ...route, file, errors };
}

async function main() {
  if (!existsSync(OUT)) await mkdir(OUT, { recursive: true });

  const browser = await chromium.launch({ headless: true });
  const ctx = await browser.newContext({ viewport: VIEWPORT, locale: "es-MX" });
  const page = await ctx.newPage();

  const results = [];
  try {
    await login(page);
    for (const r of ROUTES) {
      console.log(`→ ${r.path}`);
      results.push(await capture(page, r));
    }
  } finally {
    await browser.close();
  }

  const report = {
    generatedAt: new Date().toISOString(),
    base: BASE,
    viewport: VIEWPORT,
    routes: results.map((r) => ({ slug: r.slug, path: r.path, file: r.file, consoleErrors: r.errors.length })),
  };
  writeFileSync(path.join(OUT, "report.json"), JSON.stringify(report, null, 2));

  const md = [
    "# Reporte de consistencia visual",
    "",
    `Generado: ${report.generatedAt}  ·  Base: ${BASE}  ·  Viewport: ${VIEWPORT.width}×${VIEWPORT.height}`,
    "",
    "| # | Ruta | Errores consola | Captura |",
    "|---|------|-----------------|---------|",
    ...results.map((r, i) => `| ${i + 1} | \`${r.path}\` | ${r.errors.length} | \`${path.basename(r.file)}\` |`),
  ].join("\n");
  writeFileSync(path.join(OUT, "REPORT.md"), md);

  console.log(`\n✅ ${results.length} capturas en ${OUT}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
